### PR TITLE
Add DPTNet and WarmupStepLR scheduler

### PR DESCRIPTION
**Pull request by Emrys365**
This PR adds the [dual-path transformer network (DPTNet)](https://arxiv.org/abs/2007.13975) for time-domain speech separation, and the corresponding LR scheduler (I call it WarmupStepLR here) proposed in its paper.

### CODE REVIEW

**Comment by Emrys365** [comment]
Hi @popcornell Could you help review this PR?

File: espnet2/enh/separator/dptnet_separator.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        norm_type: str = "gLN",
+        layer: int = 6,
+        segment_size: int = 20,
+        nonlinear: str = "relu",
+    ):
+        """Dual-Path Transformer Network (DPTNet) Separator
+
+        Args:
+            input_dim: input feature dimension
+            feature_dim: intermediate feature dimension for separation
```
Comment (popcornell) [inline espnet2/enh/separator/dptnet_separator.py:37]: maybe another name than feature_dim ? bottleneck_dim ?

**Review by popcornell — COMMENTED** [review]
It looks good to me already.

File: espnet2/enh/separator/dptnet_separator.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+                'mask_spk1': torch.Tensor(Batch, Frames, Freq),
+                'mask_spk2': torch.Tensor(Batch, Frames, Freq),
+                ...
+                'mask_spkn': torch.Tensor(Batch, Frames, Freq),
+            ]
+        """
+
+        # if complex spectrum,
+        if is_complex(input):
+            feature = torch.nn.functional.relu(abs(input))
```
Comment (popcornell) [inline espnet2/enh/separator/dptnet_separator.py:115]: maybe leave relu optional ?

File: espnet2/enh/layers/dptnet.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+#
+# Ported from https://github.com/ujscjj/DPTNet
+
+import torch.nn as nn
+
+from espnet2.enh.layers.tcn import choose_norm
+from espnet.nets.pytorch_backend.nets_utils import get_activation
+
+
+class SingleTransformer(nn.Module):
```
[the remark this answers is not in the document]
Comment (Emrys365) [reply espnet2/enh/layers/dptnet.py:14]: Done.

File: espnet2/enh/separator/dptnet_separator.py (hunk 1 above)
Comment (Emrys365) [reply espnet2/enh/separator/dptnet_separator.py:37]: Oh I found that this variable is not used afterwards. So I will remove it.

File: espnet2/enh/separator/dptnet_separator.py (hunk 2 above)
Comment (Emrys365) [reply espnet2/enh/separator/dptnet_separator.py:115]: I notice that it is not necessary to apply ReLU to `abs(input)`.
For the `else` branch, I will make ReLU optional.

**Comment by Emrys365** [comment]
The trained model has been uploaded to HuggingFace: https://huggingface.co/espnet/Wangyou_Zhang_wsj0_2mix_enh_train_enh_dptnet_raw

**Comment by Emrys365** [comment]
I think this PR is ready to merge.

**Review by popcornell — APPROVED** [review]
LGTM !

**Comment by Emrys365** [comment]
Thanks! @popcornell
I will merge it.
